Guled/find another user

Somnibyte opened the pull request:
* User will now be able to view a list of users and tap on one of them.

Commit:
Add LendABookViewController.

Commit:
Add LendABookViewController. Add LendBookViewModel. Complete PL-80.

IntrepidBuildServer commented on lucien-ios/Lending/LendingTableViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+    @IBOutlet weak var nameLabel: UILabel!
+    @IBOutlet weak var emailLabel: UILabel!
+
+    override func prepareForReuse() {
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Overridden methods call super Violation: Method 'prepareForReuse()' should call to super function [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Aoverridden_super_call)

IntrepidBuildServer commented:
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

IntrepidBuildServer commented:
<!--
  0 Errors
  1 Warning: <a href="https://github.com/In...
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><a href="https://github.com/IntrepidPursuits/lucien-ios/blob/8b04b728f491c973b48ca649fbec3bfdc5785b86/LendABookViewController.swift#L120">LendABookViewController.swift#L120</a> - <strong><a href="https://github.com/IntrepidPursuits/lucien-ios/blob/8b04b728f491c973b48ca649fbec3bfdc5785b86/lucien-ios/Lending/LendABookViewController.swift#L120">lucien-ios/Lending/LendABookViewController.swift#L120</a></strong>: variable ‘self’ was written to, but never read  <br /><code>
        nextButton.rx.tap.subscribeNext { [weak self] in } &gt;&gt;&gt; disposeBag
</code></td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.491 (0.495) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.950 (6.952) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Commit:
Remove next button handler code. Add super method to prepareForReuse.

Switch must be exhaustive.

Implement Feedback.

Remove unused variables and libraries. Implement feedback.

Remove extra whitespace.

mayasaxena commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,157 @@
+            self?.tableView.reloadData()
+        }) >>> disposeBag
+
+        viewModel.users.asObservable().bind(to: tableView.rx.items(cellIdentifier: "userCell",
```
put `cellIdentifier` on its own line indented one tab from `viewModel` and adjust the other lines' indentations to match. Also put the `) {` on its own line so it's not indented so far

mayasaxena commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,157 @@
+
+    // MARK: - UITableViewDelegate
+
+    func numberOfSections(in tableView: UITableView) -> Int {
```
Don't need to implement this if there's only 1 section

mayasaxena commented on lucien-ios/Models/User.swift:
```diff
@@ -30,3 +30,14 @@ struct User: Decodable {
     }
 }
+
+struct User2: Decodable {
```
It's not good practice to name a struct the same as another with a number attached. name it more descriptively

Somnibyte replied:
That struct is never used anywhere else. It was used for testing purposes. I'll remove it. Thanks for catching that!

mayasaxena commented on lucien-ios/APIClient/LucienAPIClient.swift:
```diff
@@ -204,4 +204,29 @@ final class LucienAPIClient: APIClient {
+        let lucienRequest = LucienRequest.getAllUsers
+        let urlRequest = lucienRequest.urlRequest
+        sendRequest(urlRequest) { response in
+            print(response)
```
Remove print

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.427 (0.428) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 14.528 (14.529) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

IntrepidBuildServer commented:
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

IntrepidBuildServer commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    // MARK: - UITableViewDelegate
+    
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Trailing Whitespace Violation: Lines should not have trailing whitespace. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Atrailing_whitespace)

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_12KNSM1X1VYYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.428 (0.430) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 8.429 (8.431) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

paulrolfe commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,153 @@
+        viewModel.getAllUsers { [weak self] (users, error) in
+            if error == nil {
+                guard let users = users else { return }
+                self?.viewModel.users.value = users
```
This is a strange pattern to me. Why wouldn't the viewmodel just set this on it's own?

paulrolfe commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,153 @@
+                guard let users = users else { return }
+                self?.viewModel.users.value = users
+                DispatchQueue.main.async {
+                    self?.tableView.reloadData()
```
If you're binding the data below, do you really need this reload?

paulrolfe commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,153 @@
+                    self?.tableView.reloadData()
+                }
+            } else {
+                self?.showAlert(title: "Error", message: "Our service is currently encountering an issue. Please ensure that you are connected to the internet and try again.")
```
You could make use of the error you get here to maybe give the user more info? Maybe it's outside the scope of this work though.

Somnibyte replied:
There is no specific error for this particular endpoint. If I still have some time left over, after completing the lending feature, I'll make sure to use the errors given by the server.

paulrolfe commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,153 @@
+    // MARK: - UITableViewDelegate
+    
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 68
```
I'd like to encourage everyone to use constraints to determine cell heights these days instead of defining them in the delegate like this. https://developer.apple.com/library/content/documentation/UserExperience/Conceptual/AutolayoutPG/WorkingwithSelf-SizingTableViewCells.html

paulrolfe commented on lucien-ios/Lending/LendingBookViewModel.swift:
```diff
@@ -0,0 +1,32 @@
+
+    // MARK: - Private Constants
+
+    private let disposeBag = DisposeBag()
```
This looks like it's never used.

paulrolfe commented on lucien-ios/Lending/LendingBookViewModel.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+import RxSwift
+import RxCocoa
```
I don't see any RxCocoa work here, does it need to be imported?

IntrepidBuildServer commented on lucien-ios/Lending/LendABookViewController.swift:
```diff
@@ -0,0 +1,143 @@
+import UIKit
+import RxSwift
+
+
```
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Vertical Whitespace Violation: Limit vertical whitespace to a single empty line. Currently 2. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.intrepid.io/coding_rules#rule_key=SwiftLint%3Avertical_whitespace)

IntrepidBuildServer commented:
SonarQube analysis reported 2 issues
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

IntrepidBuildServer commented:
<!--
  0 Errors
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.368 (0.370) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.272 (6.274) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ghost commented:
SonarQube analysis reported 1 issue
* ![MINOR][MINOR] 1 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_12KNSM1X1VYYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.389 (0.391) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 8.185 (8.187) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

Commit:
Implement self-sizing table view cells.

Remove unwanted print statement.

IntrepidBuildServer commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_12KNSM1X1VYYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.368 (0.371) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 6.659 (6.661) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ghost commented:
<!--
  1 Error: fatal error: file ‘/Users/jenk...
  0 Warnings
  2 Messages: Executed 2 tests, with 0 failu...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">fatal error: file ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/Pods/Google/Headers/module.modulemap’ has been modified since the precompiled header ‘/Users/jenkins/Jenkins/workspace/s-pipeline_lucien-ios_PR-22-H3PSOV7OJ3KCVRLUTCV3H47222MCXSZ74M36WSXZCYXOA3NBTKCA/DerivedData/Build/Intermediates.noindex/PrecompiledHeaders/lucien-ios-Bridging-Header-swift_12KNSM1X1VYYJ-clang_2O9YIEWVALUCS.pch’ was built</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          2 Messages
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 2 tests, with 0 failures (0 unexpected) in 0.385 (0.387) seconds</td>
    </tr>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">Executed 1 test, with 0 failures (0 unexpected) in 48.850 (48.851) seconds</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>
